[AC-3468] feat: file message support

bang9 opened the pull request:
## Changes
- Added file message support

ticket: [AC-3468]

## Additional Notes
- figma: https://www.figma.com/design/qvfSo6jeK4t4a9jcidEjZO/2407_Support-image-%26-text-in-Widget_WIP?node-id=1-57&node-type=CANVAS&m=dev
- prd: https://sendbird.atlassian.net/wiki/spaces/AC/pages/2504884225/PRD+Image+file+support

## Checklist
Before requesting a code review, please check the following:
- [x] **[Required]** CI has passed all checks.
- [ ] **[Required]** A self-review has been conducted to ensure there are no minor mistakes.
- [ ] **[Required]** Unnecessary comments/debugging code have been removed.
- [ ] **[Required]** All requirements specified in the ticket have been accurately implemented.
- [ ] Ensure the ticket has been updated with the sprint, status, and story points.


[AC-3468]: https://sendbird.atlassian.net/browse/AC-3468?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bang9 committed:
chore: enable file upload

bang9 committed:
feat: add bot configs for replyToFile

bang9 committed:
chore: link replyToFile to uikit config

bang9 committed:
feat: implement file preview

bang9 committed:
feat: add bubble preview

bang9 committed:
feat: add file paste, update config and sdk

bang9 committed:
feat: implement drag drop

bang9 committed:
chore: rebase and apply modularization changes

bang9 committed:
chore: refactor

bang9 committed:
chore: add file viewer

bang9 committed:
feat: add file upload limit alert

bang9 committed:
chore: update gap px

chohongm commented on package.json:
```diff
@@ -82,5 +82,8 @@
-  "packageManager": "yarn@4.2.2"
+  "packageManager": "yarn@4.2.2",
+  "resolutions": {
+    "@sendbird/chat": "file:./sendbird-chat-4.14.2.tgz"
```
Is there PR for this change?

bang9 replied:
Not yet, after being deployed to specific regions, Chat SDK will be released once it has been deployed to all regions. [Branch](https://github.com/sendbird/chat-js/tree/feature/danney/v4/CLNP-4946-Add-message-property-in-FileMessage-240827)

chohongm commented on src/components/chat/hooks/useWidgetChatHandlers.ts:
```diff
@@ -14,11 +15,19 @@ export const useWidgetChatHandlers = (params: { onScrollToBottom: () => void })
       if (aiAttributesRef.current) {
-        return { ...params, data: JSON.stringify({ ai_attrs: aiAttributesRef.current }) };
+        return {
+          metaArrays: metaArray ? [metaArray] : undefined,
```
order should be
```
...params,
metaArrays: metaArray ? [metaArray] : undefined,
```

chohongm commented on src/components/chat/hooks/useWidgetChatHandlers.ts:
```diff
@@ -14,11 +15,19 @@ export const useWidgetChatHandlers = (params: { onScrollToBottom: () => void })
-        return params;
+        return {
+          metaArrays: metaArray ? [metaArray] : undefined,
+          ...params,
```
same here

chohongm commented on src/components/chat/ui/ChatInput.tsx:
```diff
@@ -28,14 +31,23 @@ export const ChatInput = () => {
+        sendUserMessage={async (params) => {
+          const processedParams = await handlers.onBeforeSendMessage(params);
+          const message = await dataSource.sendUserMessage(processedParams, () => handlers.onAfterSendMessage());
+          handlers.onAfterSendMessage();
```
It seems onAfterSendMessage being called onPending and onSuccess/onFailed. However, onAfterSendMessage() does not take message as input. How can it distinguish pending and success?

bang9 replied:
`dataSource.sendUserMessage` is a function that wraps the message handler with a promise and a callback.

```ts
function sendUserMessage(callback) {
  return new Promise((resolve, reject) => {
    channel.sendUserMessage()
      .onPending(callback)
      .onSucceeded(resolve)
      .onFailed(reject)
  });
}
```

Here’s how the logic works.

```ts
const sent = await dataSource.sendUserMessage(processedParams, () => handlers.onAfterSendMessage()); // on pending
handlers.onAfterSendMessage(); // after succeeded
```

chohongm replied:
I understood that. I meant to ask `onAfterSendMessage: () => void;` should be `onAfterSendMessage: (message: BaseMessage) => void;`. Otherwise, onAfterSendMessage will not know when it is being called (pending or success/fail).

bang9 replied:
Ah, right now it's only handling scrolling. That's why I didn't set it up to receive messages via a callback. It should be fine to add that later if needed.

chohongm commented on src/components/chat/ui/ChatInput.tsx:
```diff
@@ -28,14 +31,23 @@ export const ChatInput = () => {
+        sendFileMessage={async (params) => {
+          const processedParams = await handlers.onBeforeSendMessage(params);
+          const message = await dataSource.sendFileMessage(processedParams, () => handlers.onAfterSendMessage());
+          handlers.onAfterSendMessage();
```
same here

chohongm reviewed (commented):
<img width="794" alt="Screenshot 2024-09-10 at 4 11 34 PM" src="https://github.com/user-attachments/assets/b1df5c5b-6a47-4653-8c38-794816c53b84">

요 기능은 PR에 없는거 같은데 포함 될 예정일까요?

bang9 commented:
> 요 기능은 PR에 없는거 같은데 포함 될 예정일까요?

캡쳐된 스크린샷에 기능이 좀 많아서요! 어떤거 말씀하시는걸까요?! 파일 프리뷰 말씀하시는걸까여?

bang9 committed:
chore: apply review

chohongm commented:
현재는 파일 selection 에서 선택하면 바로 file message 가 전송이 되는데 제가 찍은 스샷으로 이해하면 select 를 해도 바로 보내지지 않고 메세지 인풋창에 선택된 파일이 노출이 되고 text 메세지도 같이 보낼수 있는 ux 로 이해를 했습니다.

bang9 commented:
> 현재는 파일 selection 에서 선택하면 바로 file message 가 전송이 되는데 제가 찍은 스샷으로 이해하면 select 를 해도 바로 보내지지 않고 메세지 인풋창에 선택된 파일이 노출이 되고 text 메세지도 같이 보낼수 있는 ux 로 이해를 했습니다.

네 맞습니다. 해당 기능은 PR 에 포함되어있습니다. 구현 자체는 UIKit 쪽 MessageInput 에 되어 있습니다.

chohongm commented:
[packages/uikit](https://github.com/sendbird/chat-ai-widget/pull/355/files#diff-6294cd105e3153d3da85f3de69680815ea331569050b72f87ec5b377c3f8cc0e) 브랜치는 따로 없을까요? feat/file-support 브랜치의 chat-ai-widget/packages/uikit 경로는 ai-widget/experimental 으로 checkout 이 되있네요.

bang9 commented:
> [packages/uikit](https://github.com/sendbird/chat-ai-widget/pull/355/files#diff-6294cd105e3153d3da85f3de69680815ea331569050b72f87ec5b377c3f8cc0e) 브랜치는 따로 없을까요? feat/file-support 브랜치의 chat-ai-widget/packages/uikit 경로는 ai-widget/experimental 으로 checkout 이 되있네요.

https://github.com/sendbird/sendbird-uikit-react/compare/acc071328e687dc2fa69c5061699f3d3548fafcf...db12d6908c34106114cfe4ee5b070348ef73a26b#diff-92b0475a052f7ab7868cf176ae3f55a7d6f7c249d22688ea708f891b1e608e76
 에서 보실 수 있습니다.

chohongm commented:
@bang9 넵 봤어요. 저는 로컬에서 직접 테스트해보고 싶은데 branch 만들어주실 수 있나요?

bang9 commented:
> @bang9 넵 봤어요. 저는 로컬에서 직접 테스트해보고 싶은데 branch 만들어주실 수 있나요?

`ai-widget/file-support` 브랜치 입니다. 리암의 로컬에서 서브모듈 헤드를 요기로 옮기시면 될거같아보입니당

chohongm reviewed (approved):
LGTM

bang9 committed:
[AC-3788] chore: file support design qa (#357)

## Changes
- Design QA changes

ticket: [AC-3788]

## Additional Notes
- 

## Checklist
Before requesting a code review, please check the following:
- [x] **[Required]** CI has passed all checks.
- [x] **[Required]** A self-review has been conducted to ensure there
are no minor mistakes.
- [x] **[Required]** Unnecessary comments/debugging code have been
removed.
- [x] **[Required]** All requirements specified in the ticket have been
accurately implemented.
- [ ] Ensure the ticket has been updated with the sprint, status, and
story points.


[AC-3788]:
https://sendbird.atlassian.net/browse/AC-3788?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
